### PR TITLE
[SERVING] Add Conv Template and Function Calling support to JSON FFI

### PR DESCRIPTION
**Pull request by Kartik14**
This PR adds conv template support to the JSON FFI Engine. Also add function calling and pass stop str to generation config.

Worked on collaboratively with @shreygupta2809 

cc @tqchen

### CODE REVIEW

File: cpp/json_ffi/conv_template.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    TOOL,
+    FUNCTION
+};
+
+MessagePlaceholders message_placeholder_from_string(const std::string& role);
+
+namespace mlc{
+namespace llm{
+namespace json_ffi{
+struct Conversation {
```
Comment (MasterJH5574) [inline cpp/json_ffi/conv_template.h:30]: Thank you @Kartik14 and @shreygupta2809 for the push! Good to have some docstrings for this class and member functions for readability.

File: cpp/json_ffi/conv_template.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    static std::optional<Conversation> FromJSON(const picojson::object& json, std::string* err);
+    static std::optional<Conversation> FromJSON(const std::string& json_str, std::string* err);
+};
+
+}
+}
+}
+
+#endif /* MLC_LLM_JSON_FFI_OPENAI_API_CONV_TEMPLATE_H */
```
Comment (MasterJH5574) [inline cpp/json_ffi/conv_template.h:61]: Need an end-of-file empty line.

File: cpp/json_ffi/openai_api_protocol.h [hunk 3]
```diff
@@ -158,11 +167,11 @@ class ChatCompletionStreamResponse {
...
-  picojson::object ToJSON();
+  picojson::object ToJSON() const;
 };
 
 }  // namespace json_ffi
 }  // namespace llm
 }  // namespace mlc
 
-#endif  // MLC_LLM_JSON_FFI_OPENAI_API_PROTOCOL_H
+#endif  // MLC_LLM_JSON_FFI_OPENAI_API_PROTOCOL_H
```
Comment (MasterJH5574) [inline cpp/json_ffi/openai_api_protocol.h:177]: Ditto here.

File: cpp/json_ffi/conv_template.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
+#ifndef MLC_LLM_JSON_FFI_OPENAI_API_CONV_TEMPLATE_H
+#define MLC_LLM_JSON_FFI_OPENAI_API_CONV_TEMPLATE_H
```
Comment (MasterJH5574) [inline cpp/json_ffi/conv_template.h:2]: `MLC_LLM_JSON_FFI_CONV_TEMPLATE_H`

File: cpp/json_ffi/conv_template.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+enum class MessagePlaceholders {
+    SYSTEM,
+    USER,
+    ASSISTANT,
+    TOOL,
+    FUNCTION
+};
+
+MessagePlaceholders message_placeholder_from_string(const std::string& role);
```
Comment (MasterJH5574) [inline cpp/json_ffi/conv_template.h:25]: Move these into the `json_ffi` namespace.

File: cpp/json_ffi/conv_template.cc [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+
+std::map<MessagePlaceholders, std::string> PLACEHOLDERS = {
+    {MessagePlaceholders::SYSTEM, "{system_message}"},
+    {MessagePlaceholders::USER, "{user_message}"},
+    {MessagePlaceholders::ASSISTANT, "{assistant_message}"},
+    {MessagePlaceholders::TOOL, "{tool_message}"},
+    {MessagePlaceholders::FUNCTION, "{function_string}"}
+};
+
+MessagePlaceholders message_placeholder_from_string(const std::string& role) {
```
Comment (MasterJH5574) [inline cpp/json_ffi/conv_template.cc:16]: Camel case for C++ functions (the same applies other functions introduced in this PR).

File: cpp/json_ffi/conv_template.cc [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    static const std::unordered_map<std::string, MessagePlaceholders> enumMap = {
```
Comment (MasterJH5574) [inline cpp/json_ffi/conv_template.cc:17]: Snake case for variables. See google c++ style guide https://google.github.io/styleguide/cppguide.html

File: cpp/json_ffi/conv_template.h [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    std::string role_content_sep;
+    std::string role_empty_sep;
+    std::vector<std::string> stop_str;
+    std::vector<int> stop_token_ids;
+    std::optional<std::string> function_string = std::nullopt;
+    std::optional<bool> use_function_calling = false;
+
+    Conversation();
+
+    static std::vector<std::string> check_message_seps(std::vector<std::string> seps);
```
Comment (MasterJH5574) [inline cpp/json_ffi/conv_template.h:49]: `const std::vector<std::string>& seps`

File: cpp/json_ffi/conv_template.cc [hunk 9]
```diff
@@ -0,0 +1,310 @@
+#include "conv_template.h"
```
Comment (MasterJH5574) [inline cpp/json_ffi/conv_template.cc:1]: We use 2 spaces as indent in C++.

File: cpp/json_ffi/openai_api_protocol.cc [hunk 10]
```diff
@@ -221,4 +466,4 @@ picojson::object ChatCompletionStreamResponse::ToJSON() {
 
 }  // namespace json_ffi
 }  // namespace llm
-}  // namespace mlc
+}  // namespace mlc
```
Comment (MasterJH5574) [inline cpp/json_ffi/openai_api_protocol.cc:469]: Ditto

**Comment by Kartik14** [comment]
@MasterJH5574 Thanks for the review! we have made the requested changes.

File: cpp/json_ffi/json_ffi_engine.cc [hunk 11]
```diff
@@ -52,32 +52,42 @@ bool JSONFFIEngine::AddRequest(std::string request_json_str, std::string request
 
   // inputs
   // TODO: Apply conv template
-  Array<Data> inputs;
+  Conversation conv_template = this->conv_template_;
+  std::vector<std::pair<std::string,
+                        std::optional<std::vector<std::unordered_map<std::string, std::string>>>>>
+      messages;
```
Comment (MasterJH5574) [inline cpp/json_ffi/json_ffi_engine.cc:58]: Given the `std::pair<....>` is too complicated here, let's introduce a class Message which has the first field `std::string` and the second field `std::optional<...>`. This is mainly for readability.

**Review by MasterJH5574 — COMMENTED** [review]
A few more nits. After these we are good to merge I think.

**Comment by MasterJH5574** [comment]
Just noticed that #2189 was merged earlier today. Could you please rebase this PR and resolve the conflicts?

**Review by shreygupta2809 — APPROVED** [review]

**Comment by shreygupta2809** [comment]
@MasterJH5574 Thanks for the review! we have made the requested changes.

**Review by MasterJH5574 — APPROVED** [review]

**Pull request merged**